scipy/signal/_spectral_py.py

```diff
@@ -85,71 +115,212 @@ def lombscargle(x,
-    freqs = np.ascontiguousarray(freqs, dtype=np.float64)
-
-    assert x.ndim == 1
-    assert y.ndim == 1
-    assert freqs.ndim == 1
 
-    if precenter:
-        pgram = _lombscargle(x, y - y.mean(), freqs)
+    # if no weights are provided, assume all data points are equally important
+    if weights is None:
+        weights = np.ones_like(x, dtype=np.float64)
+
+    # validate input shapes and if they are real dtypes
+    if x.ndim != 1:
+        raise ValueError("x array is not 1D")
+    if y.ndim != 1:
+        raise ValueError("y array is not 1D")
+    if freqs.ndim != 1:
+        raise ValueError("freqs array is not 1D")
+    if weights.ndim != 1:
+        raise ValueError("weights array is not 1D")
+    if x.shape != y.shape or x.shape != weights.shape:
+        raise ValueError("Input arrays do not have the same shape.")
+    if np.iscomplexobj(x):
+        raise TypeError("x array is a complex dtype")
+    if np.iscomplexobj(y):
+        raise TypeError("y array is a complex dtype")
+    if np.iscomplexobj(freqs):
+        raise TypeError("freqs array is a complex dtype")
+    if np.iscomplexobj(weights):
+        raise TypeError("weights array is a complex dtype")
+    
+    # check for weights summing to 0
+    if weights.sum() == 0:
+        raise ZeroDivisionError("Weights sum to 0.")
+    
+    # check for any negative weights
+    if np.any(weights < 0):
+        raise ValueError("Each weight must be >= 0.")
+
+    # validate normalize parameter
+    if isinstance(normalize, bool):
+        # if bool, convert to str literal
+        normalize = "normalize" if normalize else "power"
     else:
-        pgram = _lombscargle(x, y, freqs)
-
-    if normalize:
-        pgram *= 2 / np.dot(y, y)
+        # if neither a bool or str
+        if not isinstance(normalize, str):
+            raise TypeError("Normalize type must a bool or str.")
+
+    if normalize not in ["power", "normalize", "amplitude"]:
+        raise ValueError(
+            "Normalize must be: "
+            "False (or 'power'), "
+            "True (or 'normalize'), "
+            "or 'amplitude'."
+        )
 
+    # if input arrays are not float64, convert them
+    if x.dtype != np.float64:
+        x = x.astype(np.float64)
+    if y.dtype != np.float64:
+        y = y.astype(np.float64)
+    if freqs.dtype != np.float64:
+        freqs = freqs.astype(np.float64)
+    if weights.dtype != np.float64:
+        weights = weights.astype(np.float64)
+
+    # weight vector must sum to 1
+    weights = weights / weights.sum()
+
+    # pre-allocate arrays for intermediate and final calculations
+    coswt = np.empty_like(x)
+    sinwt = np.empty_like(x)
+    pgram = np.empty_like(freqs)
+    # store a and b so that phase can be calculated outside loop, if necessary
+    a = np.empty_like(freqs)
+    b = np.empty_like(freqs)
+
+    # perform precenter, only when called for and floating_mean is False
+    if precenter and (not floating_mean):
+        y -= y.mean()
```

Comment:
This might be a bit of bike-shedding: For more clarity, I think `precenter` should always subtract the mean, since it might produce different result together with `floating_mean`, when the input samples are not distributed well. 
Alternatively, the parameter `precenter` could be changed to the type `bool | Literal['subtract_mean', 'floating_mean']` and the parameter `floating_mean` would be removed—this would make it clearer that only one or the other method is allowed.

(Leaving it as is would be ok as well).

Reply:
I don't think it's possible, other than an infinitesimal numerical difference, that removing the mean ahead of time would produce a different result for the `floating_mean`=True condition. Which, I guess, in that case, means I could just explicitly explain the futility in the doc string, and remove the `(not floating_mean)` part of the condition. 

I don't like the idea of combining the parameters, or really of -not- making the "floating mean" (here I'm referring to the technique, not the parameter name) the default. I could be wrong, but I think that in the service of backwards compatibility, that you'd like the "base" API remain unchanged wrt to the parameter names and their defaults. And I think what you're proposing would mean that to use the "floating mean" algorithm would require the user to proactively request it. I think that would be a mistake.

Reply:
To highlight the changes to the doc string:

    Parameters
    precenter : bool, optional
        Pre-center measurement values by subtracting the mean, if True. This is
        unnecessary if `floating_mean` is True (default).

    Notes
    The algorithm used will account for any unknown offset by default, unless
    floating_mean is False. If precenter is True, it performs the operation
    ``y -= y.mean()``. However, precenter is a legacy parameter, and unnecessary 
    when floating_mean is True. Furthermore, the mean removed by precenter does not
    account for sample weights.

Reply:
Done. Unless there are other suggestions.